File: evaluator/024.metric-recorder-buffered-flush/tests/test_recorder_lifecycle.cc
#include <sstream>
#include <vector>

#include "buffered_metric_recorder.h"
#include "console_metric_recorder.h"
#include "gtest/gtest.h"
#include "metric_collector.h"
#include "metric_recorder.h"

namespace {

using nitr::case024::BufferedMetricRecorder;
using nitr::case024::ConsoleMetricRecorder;
using nitr::case024::Metric;
using nitr::case024::MetricCollector;
using nitr::case024::MetricRecorder;

TEST(ConsoleMetricRecorderTest, RecordWritesEachMetricImmediately) {
  std::ostringstream out;
  ConsoleMetricRecorder recorder(out);

  recorder.Record({"requests", 1.0});
  EXPECT_FALSE(out.str().empty())
      << "ConsoleMetricRecorder must write metrics immediately, "
         "preserving its previous behavior.";

  recorder.Record({"errors", 2.0});
  const std::string output = out.str();
  EXPECT_NE(output.find("requests"), std::string::npos);
  EXPECT_NE(output.find("errors"), std::string::npos);
}

TEST(BufferedMetricRecorderTest, DefersVisibilityBeforeCapacityOrFlush) {
  std::ostringstream out;
  BufferedMetricRecorder recorder(out, /*capacity=*/8);

  recorder.Record({"a", 1.0});
  recorder.Record({"b", 2.0});
  recorder.Record({"c", 3.0});

  EXPECT_TRUE(out.str().empty())
      << "BufferedMetricRecorder must NOT write to the underlying stream "
         "until capacity is reached or Flush is called.";
}

TEST(BufferedMetricRecorderTest, FlushesWhenCapacityIsReached) {
  std::ostringstream out;
  BufferedMetricRecorder recorder(out, /*capacity=*/3);

  recorder.Record({"a", 1.0});
  recorder.Record({"b", 2.0});
  EXPECT_TRUE(out.str().empty()) << "Buffer should still hold first two metrics.";

  recorder.Record({"c", 3.0});
  const std::string output = out.str();
  EXPECT_FALSE(output.empty())
      << "Buffer should have flushed once capacity (3) was reached.";
  EXPECT_NE(output.find('a'), std::string::npos);
  EXPECT_NE(output.find('b'), std::string::npos);
  EXPECT_NE(output.find('c'), std::string::npos);
}

TEST(BufferedMetricRecorderTest, ExplicitFlushMakesQueuedMetricsVisible) {
  std::ostringstream out;
  BufferedMetricRecorder recorder(out, /*capacity=*/100);

  recorder.Record({"queued1", 10.0});
  recorder.Record({"queued2", 20.0});
  ASSERT_TRUE(out.str().empty());

  recorder.Flush();

  const std::string output = out.str();
  EXPECT_NE(output.find("queued1"), std::string::npos);
  EXPECT_NE(output.find("queued2"), std::string::npos);
}

// The keystone test: forces the polymorphic checkpoint path.
// The collector holds a MetricRecorder& reference and must be able to flush
// any concrete recorder without knowing the concrete type.
TEST(MetricCollectorTest, CheckpointFlushesUnderlyingBufferedRecorder) {
  std::ostringstream out;
  BufferedMetricRecorder recorder(out, /*capacity=*/100);

  // Wire the collector through the abstract base reference. This is the
  // production calling convention: callers do not know the concrete recorder.
  MetricRecorder& abstract_ref = recorder;
  MetricCollector collector(abstract_ref);

  const std::vector<double> samples = {1.0, 2.0, 3.0};
  collector.Collect(samples);
  EXPECT_TRUE(out.str().empty())
      << "Buffered recorder should still be holding metrics before checkpoint.";

  collector.Checkpoint();

  const std::string output = out.str();
  EXPECT_NE(output.find("samples.count"), std::string::npos)
      << "Checkpoint must make queued metrics visible through the abstract "
         "recorder reference, without the collector knowing the concrete type.";
  EXPECT_NE(output.find("samples.mean"), std::string::npos);
  EXPECT_NE(output.find("samples.max"), std::string::npos);
}

TEST(MetricCollectorTest, CheckpointIsHarmlessForImmediateRecorder) {
  std::ostringstream out;
  ConsoleMetricRecorder recorder(out);
  MetricCollector collector(recorder);

  collector.Collect({4.0, 5.0});
  const std::string before_checkpoint = out.str();
  EXPECT_FALSE(before_checkpoint.empty())
      << "Console recorder should already have published metrics during Collect.";

  // Checkpoint on an immediate-write recorder must be safe and idempotent.
  // It should not throw, must not corrupt prior output, and need not add
  // any new content.
  collector.Checkpoint();

  const std::string after_checkpoint = out.str();
  EXPECT_GE(after_checkpoint.size(), before_checkpoint.size());
  EXPECT_NE(after_checkpoint.find("samples.count"), std::string::npos);
}

}  // namespace

File: evaluator/032.metric-recorder-buffered-flush-python/tests/test_recorder_lifecycle.py
#!/usr/bin/env python3

from __future__ import annotations

import argparse
import io
import sys
import unittest
from pathlib import Path

ARG_PARSER = argparse.ArgumentParser()
ARG_PARSER.add_argument("--case_root", required=True)
ARGS = ARG_PARSER.parse_args()

CASE_ROOT = Path(ARGS.case_root).resolve()
if str(CASE_ROOT) not in sys.path:
    sys.path.insert(0, str(CASE_ROOT))

from src.console_metric_recorder import ConsoleMetricRecorder
from src.metric_collector import MetricCollector
from src.metric_recorder import Metric, MetricRecorder

try:
    from src.buffered_metric_recorder import BufferedMetricRecorder
except ModuleNotFoundError:
    BufferedMetricRecorder = None


class RecorderLifecycleTests(unittest.TestCase):
    def test_console_recorder_writes_each_metric_immediately(self) -> None:
        out = io.StringIO()
        recorder = ConsoleMetricRecorder(out)

        recorder.record(Metric("requests", 1.0))
        self.assertNotEqual(out.getvalue(), "")

        recorder.record(Metric("errors", 2.0))
        output = out.getvalue()
        self.assertIn("requests", output)
        self.assertIn("errors", output)

    def test_buffered_recorder_class_exists(self) -> None:
        self.assertIsNotNone(
            BufferedMetricRecorder,
            msg="src.buffered_metric_recorder.BufferedMetricRecorder must exist.",
        )

    def test_buffered_recorder_defers_visibility_before_capacity_or_flush(self) -> None:
        self.assertIsNotNone(BufferedMetricRecorder)
        out = io.StringIO()
        recorder = BufferedMetricRecorder(out, capacity=8)

        recorder.record(Metric("a", 1.0))
        recorder.record(Metric("b", 2.0))
        recorder.record(Metric("c", 3.0))

        self.assertEqual(out.getvalue(), "")

    def test_buffered_recorder_flushes_when_capacity_is_reached(self) -> None:
        self.assertIsNotNone(BufferedMetricRecorder)
        out = io.StringIO()
        recorder = BufferedMetricRecorder(out, capacity=3)

        recorder.record(Metric("a", 1.0))
        recorder.record(Metric("b", 2.0))
        self.assertEqual(out.getvalue(), "")

        recorder.record(Metric("c", 3.0))
        output = out.getvalue()
        self.assertNotEqual(output, "")
        self.assertIn("a", output)
        self.assertIn("b", output)
        self.assertIn("c", output)

    def test_explicit_flush_makes_queued_metrics_visible(self) -> None:
        self.assertIsNotNone(BufferedMetricRecorder)
        out = io.StringIO()
        recorder = BufferedMetricRecorder(out, capacity=100)

        base: MetricRecorder = recorder
        base.record(Metric("queued1", 10.0))
        base.record(Metric("queued2", 20.0))
        self.assertEqual(out.getvalue(), "")

        base.flush()

        output = out.getvalue()
        self.assertIn("queued1", output)
        self.assertIn("queued2", output)

    def test_checkpoint_flushes_underlying_buffered_recorder(self) -> None:
        self.assertIsNotNone(BufferedMetricRecorder)
        out = io.StringIO()
        recorder = BufferedMetricRecorder(out, capacity=100)

        abstract_ref: MetricRecorder = recorder
        collector = MetricCollector(abstract_ref)

        collector.collect([1.0, 2.0, 3.0])
        self.assertEqual(out.getvalue(), "")

        collector.checkpoint()

        output = out.getvalue()
        self.assertIn("samples.count", output)
        self.assertIn("samples.mean", output)
        self.assertIn("samples.max", output)

    def test_checkpoint_is_harmless_for_immediate_recorder(self) -> None:
        out = io.StringIO()
        recorder = ConsoleMetricRecorder(out)
        collector = MetricCollector(recorder)

        collector.collect([4.0, 5.0])
        before_checkpoint = out.getvalue()
        self.assertNotEqual(before_checkpoint, "")

        collector.checkpoint()

        after_checkpoint = out.getvalue()
        self.assertGreaterEqual(len(after_checkpoint), len(before_checkpoint))
        self.assertIn("samples.count", after_checkpoint)


if __name__ == "__main__":
    unittest.main(argv=[sys.argv[0]])
